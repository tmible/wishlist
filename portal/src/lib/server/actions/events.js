/** @module События действий */

/**
 * Событие добавления действия
 * @constant {string}
 */
export const AddAction = {};
