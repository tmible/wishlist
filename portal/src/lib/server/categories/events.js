/** @module События категорий */

/**
 * Событие получения категорий
 * @constant {string}
 */
export const GetCategories = 'get categories';

/**
 * Событие добавления категории
 * @constant {string}
 */
export const AddCategory = 'add category';

/**
 * Событие обновления категории
 * @constant {string}
 */
export const UpdateCategory = 'update category';

/**
 * Событие удаления категории
 * @constant {string}
 */
export const DeleteCategory = 'delete category';
