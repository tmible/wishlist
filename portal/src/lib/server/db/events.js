/** @module События адаптера БД */

/**
 * Событие выполнения транзакции
 * @constant {string}
 */
export const RunTransaction = 'run transaction';

/**
 * Событие выполнения SQL выражения с проверкой авторизации
 * @constant {string}
 */
export const RunStatementAuthorized = 'run statement authorized';
