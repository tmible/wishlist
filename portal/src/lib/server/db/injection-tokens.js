/** @module Токены внедрения адаптера БД */

/**
 * Токен внедрения БД
 * @constant {string}
 */
export const Database = 'wishlist database';

/**
 * SQL выражение для посчёта изменений
 * @constant {string}
 */
export const ChangesStatement = 'changes statement';
