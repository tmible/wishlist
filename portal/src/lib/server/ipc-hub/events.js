/** @module События IPC хаба */

/**
 * Событие получения новых данных из сокета
 * @constant {{}}
 */
export const SocketData = {};

/**
 * Событие ошибки сокета
 * @constant {{}}
 */
export const SocketError = {};

/**
 * Событие проверки статуса сокета
 * @constant {{}}
 */
export const CheckSocketConnection = {};

/**
 * Событие отправки данных сокету
 * @constant {{}}
 */
export const WriteToSocket = {};
