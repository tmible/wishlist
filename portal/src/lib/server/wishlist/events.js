/** @module События списка желаний */

/**
 * Событие получения списка
 * @constant {string}
 */
export const GetWishlist = 'get wishlist';

/**
 * Событие добавления элемента к списку
 * @constant {string}
 */
export const AddItem = 'add item';

/**
 * Событие сохранения элементов разметки описания
 * @constant {string}
 */
export const InsertDescriptionEntities = 'insert description entities';

/**
 * Событие получения элемента
 * @constant {string}
 */
export const GetItem = 'get item';

/**
 * Событие получения функции, подсчитывающей количество элементов списка желаний,
 * принадлежащих пользователю, среди указанных
 * @constant {string}
 */
export const GetWishlistItemsCounter = 'get wishlist items counter';

/**
 * Событие переупорядочивания списка
 * @constant {string}
 */
export const ReorderWishlist = 'reorder wishlist';

/**
 * Событие удаления нескольких элементов из списка
 * @constant {string}
 */
export const DeleteItems = 'delete items';

/**
 * Событие обновления элемента списка
 * @constant {string}
 */
export const UpdateItem = 'update item';

/**
 * Событие удаления элементов разметки описания
 * @constant {string}
 */
export const DeleteDescriptionEntities = 'delete description entities';

/**
 * Событие удаления одного элемента из списка
 * @constant {string}
 */
export const DeleteItem = 'delete item';
