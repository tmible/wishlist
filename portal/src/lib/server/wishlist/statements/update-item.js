import { inject } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import { Database } from '$lib/server/db/injection-tokens.js';
import { UpdateItem } from '../events.js';

/** @typedef {import('$lib/wishlist/domain.js').OwnWishlistItemFormData} OwnWishlistItemFormData */

/**
 * Отображение названий атрибутов элемента списка из запроса в названия атрибутов в БД.
 * Не упомянутые совпадают
 * @constant {Map<string, string>}
 */
const LIST_ITEM_PROPERTIES_TO_DB_COLUMNS = new Map([
  [ 'categoryId', 'category_id' ],
]);

/**
 * Создание SQL выражения для обновления элемента списка желаний
 * @function updateItem
 * @param {number} userid Идентифкатор пользователя — владельца списка
 * @param {number} id Идентификатор обновляемого элемента списка
 * @param {(keyof OwnWishlistItemFormData)[]} keysToUpdate Названия атрибутов с изменениями
 * @param {Partial<OwnWishlistItemFormData>} patch Объект с изменениями
 * @returns {{ changes: number }} Количество обновлённых элементов списка
 */
const updateItem = (userid, id, keysToUpdate, patch) => {
  const keysToUpdateFiltered = keysToUpdate.filter((key) => key !== 'descriptionEntities');

  if (keysToUpdateFiltered.length <= 0) {
    return { changes: 0 };
  }

  const updatePairs = keysToUpdateFiltered.map(
    (key) => {
      const column = LIST_ITEM_PROPERTIES_TO_DB_COLUMNS.get(key) ?? key;
      const value = patch[key] === null ? patch[key] : `'${patch[key]}'`;
      return `${column} = ${value}`;
    },
  );

  return inject(Database).prepare(
    `UPDATE list SET ${updatePairs.join(', ')} WHERE id = ? AND userid = ?`,
  ).run(
    id,
    userid,
  );
};

/**
 * Подписка [выполнения SQL выражения]{@link updateItem} на соответствующее событие
 * @function initUpdateItemStatement
 * @returns {void}
 */
export const initUpdateItemStatement = () => {
  subscribe(UpdateItem, updateItem);
};
