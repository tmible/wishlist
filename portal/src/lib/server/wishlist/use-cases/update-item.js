import { inject } from '@tmible/wishlist-common/dependency-injector';
import { emit } from '@tmible/wishlist-common/event-bus';
import {
  LACK_OF_DATA_ERROR_MESSAGE,
} from '$lib/server/constants/lack-of-data-error-message.const.js';
import { RunStatementAuthorized, RunTransaction } from '$lib/server/db/events.js';
import { IPCHub } from '$lib/server/ipc-hub/injection-tokens.js';
import { DeleteDescriptionEntities, InsertDescriptionEntities, UpdateItem } from '../events.js';

/** @typedef {import('$lib/wishlist/domain.js').OwnWishlistItemFormData} OwnWishlistItemFormData */

/** @module Сценарий обновления элемента списка желаний */

/**
 * Названия атрибутов элемента списка. Используется
 * для отделения данных от метаданных в теле запроса
 * @constant {Set<string>}
 */
const LIST_ITEM_PROPERTIES = new Set([
  'name',
  'description',
  'descriptionEntities',
  'order',
  'categoryId',
]);

/**
 * Обновление элемента списка желаний
 * @function updateItem
 * @param {number} userid Идентифкатор пользователя — владельца списка
 * @param {number} id Идентификатор обновляемого элемента списка
 * @param {Partial<OwnWishlistItemFormData>} patch Объект с изменениями
 * @returns {void}
 * @throws {Error} Ошибка в случае отсутствия изменений
 */
export const updateItem = (userid, id, patch) => {
  const keysToUpdate = Object.keys(patch).filter((key) => LIST_ITEM_PROPERTIES.has(key));

  if (keysToUpdate.length <= 0) {
    throw new Error(LACK_OF_DATA_ERROR_MESSAGE);
  }

  emit(
    RunTransaction,
    () => {
      if (keysToUpdate.some((key) => key !== 'descriptionEntities')) {
        emit(RunStatementAuthorized, () => emit(UpdateItem, userid, id, keysToUpdate, patch), 1);
      }
      if (keysToUpdate.includes('descriptionEntities')) {
        emit(DeleteDescriptionEntities, id);
        emit(InsertDescriptionEntities, id, patch.descriptionEntities ?? []);
      }
    },
  );

  inject(IPCHub).sendMessage(`update ${userid}`);
};
