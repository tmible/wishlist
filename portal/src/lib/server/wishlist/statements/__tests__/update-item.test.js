import { inject } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { Database } from '$lib/server/db/injection-tokens.js';
import { UpdateItem } from '../../events.js';
import { initUpdateItemStatement } from '../update-item.js';

vi.mock('@tmible/wishlist-common/dependency-injector');
vi.mock('@tmible/wishlist-common/event-bus');

describe('wishlist / statements / update item', () => {
  let db;
  let statement;

  beforeEach(() => {
    statement = { run: vi.fn() };
    db = { prepare: vi.fn(() => statement) };
    vi.mocked(inject).mockReturnValue(db);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should subscribe to event', () => {
    initUpdateItemStatement();
    expect(vi.mocked(subscribe)).toHaveBeenCalledWith(UpdateItem, expect.any(Function));
  });

  describe('subscriber', () => {
    let eventHandler;

    beforeEach(() => {
      vi.mocked(subscribe).mockImplementationOnce((event, handler) => eventHandler = handler);
      initUpdateItemStatement();
    });

    it('should return if patch is empty', () => {
      expect(eventHandler('userid', 'id', [ 'descriptionEntities' ], {})).toEqual({ changes: 0 });
    });

    it('should inject database', () => {
      eventHandler('userid', 'id', [ 'name', 'description' ], { name: 'name', description: null });
      expect(vi.mocked(inject)).toHaveBeenCalledWith(Database);
    });

    it('should prepare statement', () => {
      eventHandler('userid', 'id', [ 'name', 'description' ], { name: 'name', description: null });
      expect(db.prepare.mock.calls[0]).toMatchSnapshot();
    });

    it('should run statement on event emit', () => {
      eventHandler('userid', 'id', [ 'name', 'description' ], { name: 'name', description: null });
      expect(statement.run).toHaveBeenCalledWith('id', 'userid');
    });
  });
});
