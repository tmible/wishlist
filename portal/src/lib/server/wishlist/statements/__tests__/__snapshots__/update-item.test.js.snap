// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wishlist / statements / update item > subscriber > should prepare statement 1`] = `
[
  "UPDATE list SET name = 'name', description = null WHERE id = ? AND userid = ?",
]
`;
